database/pebble/db.go

```diff
@@ -0,0 +1,341 @@
+		BytesPerSync:                cfg.BytesPerSync,
+		Comparer:                    comparer,
+		WALBytesPerSync:             cfg.WALBytesPerSync,
+		MemTableStopWritesThreshold: cfg.MemTableStopWritesThreshold,
+		MemTableSize:                cfg.MemTableSize,
+		MaxOpenFiles:                cfg.MaxOpenFiles,
+		MaxConcurrentCompactions:    runtime.NumCPU, // TODO what should default be?
+		Levels:                      make([]pebble.LevelOptions, numLevels),
+	}
+	for i := 0; i < len(opts.Levels); i++ {
+		l := &opts.Levels[i]
+		l.TargetFileSize = targetFileSize
+		l.FilterPolicy = filterPolicy
+	}
+	opts.Experimental.ReadSamplingMultiplier = -1 // explicitly disable seek compaction
+
+	log.Info("opening pebble")
+	log.Debug("config", zap.Any("config", opts))
+
+	db, err := pebble.Open(file, opts)
+	if err != nil {
+		return nil, err
+	}
+
+	return &Database{
+		pebbleDB:      db,
+		openIterators: set.Set[*iter]{},
+	}, nil
+}
+
+func (db *Database) Close() error {
+	db.lock.Lock()
+	defer db.lock.Unlock()
+
+	if db.closed {
+		return database.ErrClosed
+	}
+
+	db.closed = true
+
+	for iter := range db.openIterators {
+		iter.lock.Lock()
+		iter.release()
+		iter.lock.Unlock()
+	}
+	db.openIterators.Clear()
+
+	return updateError(db.pebbleDB.Close())
+}
+
+func (db *Database) HealthCheck(_ context.Context) (interface{}, error) {
+	db.lock.RLock()
+	defer db.lock.RUnlock()
+
+	if db.closed {
+		return nil, database.ErrClosed
+	}
+	return nil, nil
+}
+
+func (db *Database) Has(key []byte) (bool, error) {
+	db.lock.RLock()
+	defer db.lock.RUnlock()
+
+	if db.closed {
+		return false, database.ErrClosed
+	}
+
+	_, closer, err := db.pebbleDB.Get(key)
+	if err == pebble.ErrNotFound {
+		return false, nil
+	}
+	if err != nil {
+		return false, updateError(err)
+	}
+	return true, closer.Close()
+}
+
+func (db *Database) Get(key []byte) ([]byte, error) {
+	db.lock.RLock()
+	defer db.lock.RUnlock()
+
+	if db.closed {
+		return nil, database.ErrClosed
+	}
+
+	data, closer, err := db.pebbleDB.Get(key)
+	if err != nil {
+		return nil, updateError(err)
+	}
+	return slices.Clone(data), closer.Close()
+}
+
+func (db *Database) Put(key []byte, value []byte) error {
+	db.lock.Lock()
+	defer db.lock.Unlock()
+
+	if db.closed {
+		return database.ErrClosed
+	}
+
+	return updateError(db.pebbleDB.Set(key, value, pebble.Sync))
+}
+
+func (db *Database) Delete(key []byte) error {
+	db.lock.Lock()
+	defer db.lock.Unlock()
+
+	if db.closed {
+		return database.ErrClosed
+	}
+
+	return updateError(db.pebbleDB.Delete(key, pebble.Sync))
+}
+
+func (db *Database) Compact(start []byte, limit []byte) error {
+	db.lock.Lock()
+	defer db.lock.Unlock()
+
+	switch {
+	case db.closed:
+		return database.ErrClosed
+	case comparer.Compare(start, limit) >= 0:
+		// pebble's Compact will no-op & error if start >= limit
+		// according to pebble's comparer.
+		return nil
+	case limit != nil:
+		return updateError(db.pebbleDB.Compact(start, limit, true /* parallelize */))
```

Comment:
I think this should be `false`

Reply:
Why? It looks like this method is synchronous in either case.

Reply:
If you don't agree, please make it a function arg?

This will spin up a number of goroutines to concurrently compact different ranges...some users may not want this.

Reply:
made this a config